pkg/apiserver/certificate/selfsignedcert_provider.go

```diff
@@ -0,0 +1,305 @@
+	// Set the CertKey and CertDirectory to generate the certificate files.
+	secureServing.ServerCert.CertDirectory = caConfig.SelfSignedCertDir
+	secureServing.ServerCert.CertKey.CertFile = filepath.Join(caConfig.SelfSignedCertDir, caConfig.PairName+".crt")
+	secureServing.ServerCert.CertKey.KeyFile = filepath.Join(caConfig.SelfSignedCertDir, caConfig.PairName+".key")
+
+	provider := &selfSignedCertProvider{
+		client:        client,
+		secureServing: secureServing,
+		caConfig:      caConfig,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "selfSignedCertProvider"),
+		clock:         clockutils.RealClock{},
+	}
+
+	if caConfig.TLSSecretName != "" {
+		provider.secretNamespace = env.GetAntreaNamespace()
+		provider.secretInformer = coreinformers.NewFilteredSecretInformer(client, provider.secretNamespace, 12*time.Hour,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+			func(options *metav1.ListOptions) {
+				options.FieldSelector = fields.OneTermEqualSelector("metadata.name", caConfig.TLSSecretName).String()
+			})
+		provider.secretInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { provider.enqueue() },
+			UpdateFunc: func(_, _ interface{}) { provider.enqueue() },
+			DeleteFunc: func(obj interface{}) { provider.enqueue() },
+		})
+		provider.secretLister = corelisters.NewSecretLister(provider.secretInformer.GetIndexer())
+	}
+	if err := provider.rotateSelfSignedCertificate(); err != nil {
+		return nil, err
+	}
+	return provider, nil
+}
+
+func (p *selfSignedCertProvider) RunOnce(ctx context.Context) error {
+	return p.rotateSelfSignedCertificate()
+}
+
+func (p *selfSignedCertProvider) Run(ctx context.Context, workers int) {
+	defer p.queue.ShutDown()
+
+	klog.Infof("Starting selfSignedCertProvider")
+	defer klog.Infof("Shutting down selfSignedCertProvider")
+
+	if p.secretInformer != nil {
+		go p.secretInformer.Run(ctx.Done())
+	}
+
+	// doesn't matter what workers say, only start one.
+	go wait.Until(p.runWorker, time.Second, ctx.Done())
+	// check if the certificate should be regenerated periodically.
+	go wait.Until(p.enqueue, time.Hour, ctx.Done())
+
+	<-ctx.Done()
+}
+
+func (p *selfSignedCertProvider) Name() string {
+	return "self-signed cert"
+}
+
+func (p *selfSignedCertProvider) CurrentCABundleContent() []byte {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	return p.cert
+}
+
+func (p *selfSignedCertProvider) VerifyOptions() (x509.VerifyOptions, bool) {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	if p.verifyOptions == nil {
+		return x509.VerifyOptions{}, false
+	}
+	return *p.verifyOptions, true
+}
+
+func newVerifyOptions(caBundle []byte) *x509.VerifyOptions {
+	// We don't really use the CA bundle to verify clients, this is just to follow DynamicFileCAContent.
```

Comment:
just to double-check: with this, you mean that the generated cert is never used for client-side auth, only server-side auth?

Reply:
The generated cert is used for client-side server authentication. The server, antrea-controller, doesn't validate clients' TLS using this cert and it can't use it because none of the client cert is signed by it. It uses token to validate clients.

I added the comment because it sets the key usage to "ExtKeyUsageClientAuth" which might be confusing for people who wonders how can this CA bundle be used for server-side client authentication as we never generate any client cert. It's to follow DynamicFileCAContent's implementation.The verify option will never be used actually.

Reply:
> only server-side auth

That was poor phrasing on my part. I indeed meant server authentication on the client-side...